### DISCUSSION TITLE
.lib files on Windows are missing

### DISCUSSION THREAD

**Discussion by Faster3ck** [ImageMagick/ImageMagick#7241] [Development]
### ImageMagick version

7.1.1-30

### Operating system, version and so on

Windows

### Description

Hello,

Why does ImageMagick for Windows no longer include **.lib** files for C/C++ development? I discovered, via [this git commit to the ImageMagick-Windows repository](https://github.com/ImageMagick/ImageMagick-Windows/commit/8027f997a27b70028f2deb221cebfeacccdb84e5), that as of a few days ago, these files were intentionally removed from the INNO Setup installation scripts. 
Now, the ImageMagick installers only contain the header files, without the libraries.

Is there an alternative way to install the ImageMagick .lib files on Windows?

Greetings,
Francesco

**Comment by dlemstra** [comment]
We removed them because we did not think anyone would use them. What kind of use case do you have that you would use those `.lib` files instead of the `.dll` files?

**Reply by Faster3ck** [reply]
Hello and thank you for your response.

I have developed a program that uses ImageMagick's C++ libraries to do operations on images.

From what I know, in order to properly link ImageMagick libraries to my C++ program, .lib files are required during the compile process.

My program uses cmake and without the lib files I get this error:

```
CMake Error at C:/Program Files/CMake/share/cmake-3.29/Modules/FindPackageHandleStandardArgs.cmake:230 (message):
  Could NOT find ImageMagick (missing: ImageMagick_Magick++_LIBRARY
  ImageMagick_MagickWand_LIBRARY ImageMagick_MagickCore_LIBRARY) (found
  version "7.1.1-30")
Call Stack (most recent call first):
  C:/Program Files/CMake/share/cmake-3.29/Modules/FindPackageHandleStandardArgs.cmake:600 (_FPHSA_FAILURE_MESSAGE)
  C:/Program Files/CMake/share/cmake-3.29/Modules/FindImageMagick.cmake:312 (FIND_PACKAGE_HANDLE_STANDARD_ARGS)
  CMakeLists.txt:31 (find_package)


-- Configuring incomplete, errors occurred!
```

The line that I'm using for including the IM components is the following one:
`find_package(ImageMagick REQUIRED COMPONENTS Magick++ MagickWand MagickCore)`


Is there some aspect I am not considering? Is there an alternative way to compile a project in C++ without the .lib files?
I have done several searches but have not found anything different from what is reported on the [cmake page](https://cmake.org/cmake/help/latest/module/FindImageMagick.html).

Thanks in advance!

**Reply by dlemstra** [reply]
It looks like cmake is trying to find the `.lib` files but I wonder if they are used in your program. Maybe you only need the include files?

**Reply by Faster3ck** [reply]
To compile my program, I used both qmake and CMake with Visual Studio 2022. I removed all references to the .lib libraries from the files and included only the header files. I received warnings from the linker due to missing 'dll-interfaces' and an error stating, `'fatal error LNK1120: 80 unresolved externals.'`

[In this gist](https://gist.github.com/Faster3ck/a8f398366bbfd0af36c0a46cf0b67158), I have included the complete compilation error. Here are the original [CMake file](https://github.com/Faster3ck/Converseen/blob/main/CMakeLists.txt) and [qmake file](https://github.com/Faster3ck/Converseen/blob/main/converseen.pro).

If I am not mistaken, my understanding is that without the .lib files, the compiler cannot find references to the ImageMagick DLLs.

**Reply by dlemstra** [reply]
It doesn't look like your project needs those `.lib` files. But you will need to tweak your `CMakeLists.txt` file to fix this.

**Reply by simonmurdock** [reply]
I actually have this same issue just using VS 2022, the linker is unable to find `unresolved external symbol "__declspec(dllimport) void __cdecl Magick::InitializeMagick(char const *)" (__imp_?InitializeMagick@Magick@@YAXPEBD@Z) referenced in function main`

I am a C++ noob however and not sure if this is my fault or not. Other libraries I have included to provide a .lib however.
Update - Seems to have been fixed by referencing the modules dir for the linker

**Reply by arbmarton** [reply]
To link to a .dll on Windows you need the corresponding .lib files. You should revert this commit.

**Reply by dlemstra** [reply]
There is no need to link those `.lib` files. You can reference the `.dll` files instead. The methods that you are using are exported from the `.dll` files that are provided.

**Reply by arbmarton** [reply]
Here is a nice description about how linking to dynamic libraries works on Windows using C++: https://stackoverflow.com/questions/7845886/linking-dll-in-visual-studio

With this in mind, can you please elaborate how can I use your distribution for C++ development on Windows? Do you intend people to use LoadLibrary or something like that? I'm personally not used to that kind of workflow, but if you consider yourself an expert in C++ development, please elucidate me on why that is preferable to distributing the .lib files. Thanks!

**Reply by dlemstra** [reply]
I am not sure why you think I said that I am expert in C++ development. Other people reported that they got it working using just the `.dll` files. I wasn't aware that those files are used when linking `.dll` files on Windows. I assumed they were optional and only used for static linking but thanks for pointing out they are not. We will make sure these files will be added again in the next version.
